### PR TITLE
(FACT-1778) Attempt to canonicalize search paths

### PR DESCRIPTION
**Pull request by MikaelSmith**
Attempt to canonicalize search paths, then fall back to using absolute
path. This preserves the fix for FACT-1510, but also ensures we
deduplicate most paths.

This primarily fixes that in Puppet we load `lib/facter` from both
Puppet's search paths and Facter's internal handling of Ruby's LOAD_PATH.
Previously canonicalization would ensure that the resulting file paths for
custom facts were identical, so when we check whether we've previously
loaded a file we would identify duplicates and avoid loading them again.
The fix for FACT-1510 broke that. This is the least-impact approach to
fixing it.

### CODE REVIEW

File: lib/src/ruby/module.cc [hunk 1]
```diff
@@ -953,16 +963,16 @@ namespace facter { namespace ruby {
 
         // Look for "facter" subdirectories on the load path
         for (auto const& directory : ruby.get_load_path()) {
-            // Get the absolute directory name
             boost::system::error_code ec;
-            path dir = absolute(directory);
+            auto dir = canonicalize(directory).string();
```
Comment (branan) [inline lib/src/ruby/module.cc]: I think we want to keep `dir` as a `path` here, and only make it a `string` down below? (line 979)

Comment (branan) [inline lib/src/ruby/module.cc]: Since this is now a string, you need to remove the `dir.string` below on line 979

Comment (MikaelSmith) [reply lib/src/ruby/module.cc]: Made a few more changes.

File: lib/src/ruby/module.cc [hunk 2]
```diff
@@ -953,20 +958,19 @@ namespace facter { namespace ruby {
 
         // Look for "facter" subdirectories on the load path
         for (auto const& directory : ruby.get_load_path()) {
-            // Get the absolute directory name
             boost::system::error_code ec;
-            path dir = absolute(directory);
+            // Use forward-slash to keep this consistent with Ruby conventions.
+            auto dir = canonicalize(directory) + "/facter";
```
Comment (MikaelSmith) [inline lib/src/ruby/module.cc]: I can try to add a test that adds a Ruby LOAD_PATH for a fixture, then also sets it as a search path, and make sure the custom fact is only loaded once (somehow). Detecting that is the slightly tricky part.

**Pull request merged**